memutils/memory/rmm_alloc.cu

```diff
@@ -0,0 +1,258 @@
+//  Copyright (c) 2017-2018 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cuda_runtime.h>
+#include <cuda_profiler_api.h>
+#include <rmm/rmm.h>
+#include <rmm/rmm_api.h>
+#include <cstdio>
+#include <cstring>
+
+#include "../memory.h"
+
+const int MAX_ERROR_LEN = 100;
+
+// checkCUDAError checks the cuda error of last runtime calls and returns the
+// pointer to the buffer of error message. This buffer needs to be released
+// by caller or upper callers.
+char *checkCUDAError(const char *message) {
+  cudaError_t error = cudaGetLastError();
+  if (error != cudaSuccess) {
+    char *buffer = reinterpret_cast<char *>(malloc(MAX_ERROR_LEN));
+    snprintf(buffer, MAX_ERROR_LEN,
+             "ERROR when calling CUDA functions from host: %s: %s\n",
+             message, cudaGetErrorString(error));
+    return buffer;
+  }
+  return NULL;
+}
+
+char *checkRMMError(rmmError_t rmmError, const char* message) {
+  if (rmmError != RMM_SUCCESS) {
+    char *buffer = reinterpret_cast<char *>(malloc(MAX_ERROR_LEN));
+    snprintf(buffer, MAX_ERROR_LEN,
+             "ERROR when calling RMM functions: %s: %s\n",
+             message, rmmGetErrorString(rmmError));
+    return buffer;
+  }
+  return NULL;
+}
+
+DeviceMemoryFlags GetFlags() {
+  return DEVICE_MEMORY_IMPLEMENTATION_FLAG & POOLED_MEMORY_FLAG;
+}
+
+CGoCallResHandle Init() {
+  CGoCallResHandle resHandle = GetDeviceCount();
+  if (resHandle.pStrErr != nullptr) {
+    return resHandle;
+  }
+
+  size_t deviceCount = reinterpret_cast<size_t>(resHandle.res);
+  for (size_t device = 0; device < deviceCount; device++) {
+    cudaSetDevice(device);
+    rmmOptions_t options = {
+        PoolAllocation,
+        0,  // Default to half ot total memory
+        false  // Disable logging.
+    };
+    resHandle.pStrErr = checkRMMError(rmmInitialize(&options), "rmmInitialize");
+    if (resHandle.pStrErr != nullptr) {
+      return resHandle;
+    }
+  }
+  return resHandle;
+}
+
+CGoCallResHandle DeviceAllocate(size_t bytes, int device) {
+  CGoCallResHandle resHandle = {NULL, NULL};
+  cudaSetDevice(device);
+  // For now use default stream to avoid changing the memory allocation
```

Comment:
this means deviceAlloc is still sync (for now)?

Reply:
if it's an interface, does it make sense to create an abstract class that has these methods virtual?

Reply:
nvm if the goal is to replace existing code, dont bother

Reply:
yes they are synced per device at host, but they will not cause device synchronization if the memory is already pooled.